packages/@aws-cdk/aws-cfn-include/README.md

```diff
@@ -0,0 +1,190 @@
+# Include CloudFormation templates in the CDK
+
+<!--BEGIN STABILITY BANNER-->
+
+---
+
+![Stability: Experimental](https://img.shields.io/badge/stability-Experimental-important.svg?style=for-the-badge)
+
+> **This is a _developer preview_ (public beta) module.**
+>
+> All classes with the `Cfn` prefix in this module ([CFN Resources](https://docs.aws.amazon.com/cdk/latest/guide/constructs.html#constructs_lib))
+> are auto-generated from CloudFormation. They are stable and safe to use.
+>
+> However, all other classes, i.e., higher level constructs, are under active development and subject to non-backward
+> compatible changes or removal in any future version. These are not subject to the [Semantic Versioning](https://semver.org/) model.
+> This means that while you may use them, you may need to update your source code when upgrading to a newer version of this package.
+
+---
+<!--END STABILITY BANNER-->
+
+This module contains a set of classes whose goal is to facilitate working
+with existing CloudFormation templates in the CDK.
+It can be thought of as an extension of the capabilities of the
+[`CfnInclude` class](../@aws-cdk/core/lib/cfn-include.ts).
+
+## Basic usage
+
+Let's say you have the file `my-template.json` with the following contents:
+
+```json
+{
+  "Resources": {
+    "Bucket": {
+      "Type": "AWS::S3::Bucket",
+      "Properties": {
+        "BucketName": "some-bucket-name"
+      }
+    }
+  }
+}
+```
+
+You can include it in your CDK application using the following code:
+
+```typescript
+import * as inc from '@aws-cdk/aws-cfn-include';
+
+const cfnTemplate = inc.CdkInclude.includeJsonTemplate(this, 'my-template.json');
```

Comment:
The rational is the following:

One of the biggest use cases for this feature is migrating existing CloudFormation stacks to the CDK. Because of how scoping works in the CDK, if you pass a `Stack` as `scope` to `includeJsonTemplate`, something like this:

```ts
class MyStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope. id, props);

    inc.CdkInclude.includeJsonTemplate(this, 'my-template.json');
  }
}
```

All of the logical IDs from the original template will be preserved in the resulting CDK template, which makes for a clean, incremental migration.

If `CdkInclude` was a construct though, doing this:

```ts
  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope. id, props);

    new inc.CdkInclude(this, 'MyTemplate', {
      jsonFile: 'my-template.json',
    });
  }
}
```

Would result in all logical IDs from the original template changing, and that means a deployment of this existing stack would result in a replacement of every single resource inside it, making migration pretty much impossible.

Reply:
Can we make `CdkInclude` a `Stack`?

Reply:
In theory, yes. But I don't think it's the best experience we can provide.

It would force you to write non-idiomatic code - you cannot have a separate subclass of `Stack` in the `lib` folder, like in the above example with `MyStack`. You would have to call `new CdkInlcude(app)` in your main entrypoint TS file in `bin`, and then add any other constructs to the `CdkInclude` stack that you want in that same entrypoint file. The way it is modeled now, the include part is completely encapsulated inside your `Stack` subclass, along with any other resources the stack needs, exactly like other CDK stacks today that don't use include.

Also, if `CdkInclude` was a stack, it would be impossible to merge two CloudFormation templates into one CDK stack using including, like I show in the ReadMe of the module.

Reply:
I understand the use cases but I don’t think those justify the weird API. The current `CfnInclude` preserves logical ids and is still offered as an idiomatic construct.

Also, if you call this from any scope that is not the stack itself, the logicals will not be preserved and that’s quite unintuitive and can trip people up.

How about this: add an option to `CdkInclude` which preserves logical IDs. This can be implemented by looking up the `Stack` and adding all the child nodes directly to it. So it will work from every scope.

I think we will wa t this to be disabled by default.

```ts
new CdkInclude(this, `foo`, {
  templateFile: “foo.json”,
  preserveLogical: true
});
```

Reply:
So if `preserveLogicalNames` is `true`, the `id` parameter is completely ignored? That doesn't sound right...

Reply:
I don't think it's a major issue. This is the current behavior of `CfnInclude`. All construct have IDs and sometimes they are not necessarily reflected in the synthesized output.

Reply:
Done in the newest revision - let me know if that's better!